jvm/src/main/kotlin/app/cash/trifle/Trifle.kt

```diff
@@ -15,4 +29,70 @@ object Trifle {
     EndEntityDelegate by delegate {
     constructor(keysetHandle: KeysetHandle) : this(TinkDelegate(keysetHandle))
   }
+
+  fun SignedData.verify(certAnchor: ByteArray): Boolean {
+    val TRIFLE_SIGNING_MESSAGE_VERSION: Int = 0
+
+    if (certAnchor == null || certAnchor.isEmpty()) {
+      throw Exception("Invalid input root cert")
+    }
+
+    // Validate and extract signed data
+    //   `        raw_data = raw_data,
+    //           signature = signature,
+    //           certificates = certificates,
+    val messageIn = checkNotNull(enveloped_data)
+    val signatureIn = checkNotNull(signature)
+    val certificatesIn = certificates
+
+    val signingMessage = SignedData.EnvelopedData.ADAPTER.decode(messageIn)
+
+    // Validate and extract signing message
+    //       builder.version = version
+    //       builder.signing_algorithm = signing_algorithm
+    //       builder.data_ = data_
+    check(signingMessage.version == TRIFLE_SIGNING_MESSAGE_VERSION)
+    val signingAlgorithm = when (signingMessage.signing_algorithm) {
+      SignedData.Algorithm.ECDSA_SHA256 -> TrifleAlgorithmIdentifier.ECDSASha256AlgorithmIdentifier
+      else -> {
+        throw Exception("Unsupported signing algorithm")
+      }
+    }
+
+    val cf = CertificateFactory.getInstance("X.509");
+    val root: Certificate? = cf.generateCertificate(ByteArrayInputStream(certAnchor))
+    val list = ArrayList<Certificate>()
+
+    // Set up certPath. Remove root if in certPath
```

Comment:
does this remove root or device cert? root is the last element right?

Reply:
`list.remove(root)`
removes the root.  The intention of this is in case the trust anchor is included in the SignedData. For certPathValidator to work, the trust anchor must be removed. 
A complication is if the current cert chain included the trust anchor as an intermediate cert. I can be convinced to drop all the certs after the root, since client side validates the cert chain and so we can be assured that the order of certs is correct.
@alvinsee thoughts?

Reply:
it's not really important whether or not the client does the verification (the client is considered potentially malicious), but I do think we want to enable a 'verify only up to a specified intermediate cert', in which case dropping the ones after the root also makes sense.  More concretely:

Client can pass A -> B -> C -.D

and we can call signedData.verify(B), which should work.

Reply:
I think what I do not quite get is the iterator here:
`for(i in 1..certificatesIn.size)`
which is indices inclusive at position 1 to the last element. In our case this list would be a list of size two which describes:
`[deviceCert, rootCert]`
in this case, wouldnt this just iterate on the root certificate since [1, 2)?